vertical-pod-autoscaler/enhancements/4016-in-place-updates-support/README.md

```diff
@@ -17,88 +17,135 @@
+    the container without restarting. However, it does not guarantee that a restart will not
+    happen.
+  - `RestartContainer` which indicates that any mutation to the resource requires a restart (for
+    example, this is important for Java apps using the `-xmxN` which are unable to resize memory
+    without restarting).
+
+Note that resize operations will NOT change the pod's quality of service (QoS) class.
 
 [In-place update of pod resources KEP]: https://github.com/kubernetes/enhancements/issues/1287
-[Kubernetes 1.27]: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.27.md#api-change-3
-[`ResizePolicy`]: https://github.com/kubernetes/api/blob/8360d82aecbc72aa039281a394ebed2eaf0c0ccc/core/v1/types.go#L2448
+[`/resize` subresource]:https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#api-changes
+[`ResizePolicy`]: https://github.com/kubernetes/api/blob/4dccc5e86b957cea946a63c4f052ee7dec3946ce/core/v1/types.go#L2636
 
 ## Design Details
 
-In the update modes existing before (`Initial` and `Recreate`) only admission controller was changing pod spec (updater
-was responsible for evicting pods so that admission controller could change them during admission).
+Today, only the VPA admission controller is responsible for changing the pod spec.
+
+The VPA updater is responsible for evicting pods so that the admission controller can change them
+during admission.
+
+In the newly added `InPlaceOrRecreate` mode, the VPA Updater will attempt to execute in-plac
+updates FIRST. In certain situations, if it is unable to process an in-place update in time, it
+will evict the pod to force a change.
 
-In the newly added `InPlaceOnly` and `InPlaceOrRecreate` modes VPA Updater will execute in-place updates (which change
-pod spec without involving admission controller).
+We classify three types of updates in the context of this new mode:
 
-### Applying Updates During Pod Admission
+1. Updates on pod admission
+2. Disruption-free updates
+3. Disruptive updates
 
-For VPAs in `InPlaceOnly` and `InPlaceOrRecreate` modes VPA Admission Controller will apply updates to starting pods,
-like it does for VPAs in `Initial`, `Auto`, and `Recreate` modes.
+**NOTE:** Here "disruptive" is from the perspective of VPA. As mentioned above, the container
+runtime MAY choose to restart a container or pod as needed.
 
-### Applying Disruption-free Updates
+More on these two types of changes below.
 
-When an update only changes resources for which a container indicates that it doesn't require a restart, then VPA can
-attempt to actuate the change without disrupting the pod. VPA Updater will:
+### 1. Applying Updates During Pod Admission
+
+For VPAs using the new `InPlaceOrRecreate` mode, the VPA Admission Controller will apply updates to
+starting pods just as it does for VPAs in `Initial`, `Auto`, and `Recreate` modes.
```

Comment:
I'm not entirely sure if this AEP is the correct place to make this comment, but the promise of `Auto` was that it would one day support in-place, see https://github.com/kubernetes/autoscaler/blob/78c8173b979316f892327022d53369760b000210/vertical-pod-autoscaler/docs/api.md#updatemode

We've been promising users that if they use Auto today, one day in the future the VPA will start resizing their pods for them without recreating them. 
It seems that this may not be the case since the container runtime can't guarantee it. 

Do we need to start thinking about what to do with Auto in the future?

Again, I'm unsure if this really fits into the context of this AEP, but I thought I'd bring it up.

Reply:
My 2 cents on this are that, switching `Auto` to be `InPlace` or some variation of `InPlace` would be a breaking change that should be in something like VPA 2.0. It may not "break" someone's configuration, but the change probably won't be the same expected behaviour in VPA 0.x and 1.x when the updater was freely evicting pods in `Auto` mode. Although, what I'm talking about is probably a discussion for the future timeline of VPA in general.

Reply:
Changing the behavior of Auto, good idea. Major version bump to accompany that, also a good idea.

Reply:
With the discussion around changing the behavior of `auto`, we're talking about a future where in-place vertical scaling is GA and not under a feature gate, right? IMHO it would feel a bit weird to make the default behavior of VPA depend on a beta feature gate.

Reply:
> With the discussion around changing the behavior of `auto`, we're talking about a future where in-place vertical scaling is GA and not under a feature gate, right?

Correct. Since the work on in-place is starting to happen now, and we are diverging on the promise given to users, it seems appropriate to discuss Auto now, just so it's considered in our discussions as we move forward with the in-place rollout.

Reply:
Specified in the AEP that the intention is that this eventually graduates into the `Auto` mode.